### PR TITLE
Fix optEarlyPropRewriteTree

### PR DESCRIPTION
**Pull request by sandreenko**
There were two close problems:
1. It created non-unique ids, because used `CopyFrom`.
2. It did not update side effects when `tree` was replaced or deleted. We didn't hit this issue because in most cases `tree` was replaced in-place and `gtUpdateSideEffects(stmt, tree); `  worked correct.

Simplify the algorithm using `gtCloneExpr` and always replace `tree` with a new value.

This PR is looking for a case, where we have different types.

No spmi diffs.

### CODE REVIEW

**Comment by sandreenko** [comment]
Fixes cases like that:
```
After optEarlyProp:

***** BB01, stmt 1
     ( 16, 12) [000008] ------------             *  STMT      void  (IL 0x000...0x007)
N009 ( 16, 12) [000005] --CXG-------             |  /--*  CALL help ref    HELPER.CORINFO_HELP_NEWARR_1_VC
N005 (  1,  4) [000002] ------------ arg0 in ecx |  |  +--*  CNS_INT(h) int    0x4EED480 class
N006 (  1,  1) [000001] ------------ arg1 in edx |  |  \--*  CNS_INT   int    16
N011 ( 16, 12) [000007] -ACXG---R---             \--*  ASG       ref
N010 (  1,  1) [000006] D------N----                \--*  LCL_VAR   ref    V01 loc0         d:2

********************

***** BB02, stmt 7
     (  5,  4) [000107] ------------             *  STMT      void  (IL 0x008...  ???)
N001 (  1,  1) [000001] ------------             |  /--*  CNS_INT   int    16
N003 (  5,  4) [000106] -A------R---             \--*  ASG       int
N002 (  3,  2) [000105] D------N----                \--*  LCL_VAR   int    V12 tmp10        d:2 
```
where `[000001]` was cloned and is represented twice in the list.

**Comment by sandreenko** [comment]
PTAL @dotnet/jit-contrib .
Was found by the checker (Check and fix that all tree nodes are unique #14411).

File: src/jit/earlyprop.cpp [hunk 1]
```diff
@@ -307,72 +309,57 @@ bool Compiler::optEarlyPropRewriteTree(GenTreePtr tree)
             {
                 GenTreeBoundsChk* check = tree->gtNext->AsBoundsChk();
 
-                if ((check->gtArrLen == tree) && check->gtIndex->IsCnsIntOrI() &&
-                    (check->gtIndex->AsIntCon()->IconValue() >= 0) &&
-                    (check->gtIndex->AsIntCon()->IconValue() < actualVal->AsIntCon()->IconValue()))
+                if ((check->gtArrLen == tree) && check->gtIndex->IsCnsIntOrI())
```
Comment (mikedn) [inline src/jit/earlyprop.cpp]: What are these changes for?

Comment (sandreenko) [reply src/jit/earlyprop.cpp]: To remove copy paste of `actualVal->AsIntCon()->IconValue()` and `check->gtIndex->AsIntCon()->IconValue()`.

File: src/jit/earlyprop.cpp [hunk 2]
```diff
@@ -307,72 +309,57 @@ bool Compiler::optEarlyPropRewriteTree(GenTreePtr tree)
...
-        GenTreePtr actualValCopy;
-        var_types  origType = tree->gtType;
+
+        GenTreePtr actualValClone = gtCloneExpr(actualVal);
+
+        if (actualValClone->gtType != tree->gtType)
+        {
+            assert(actualValClone->gtType == TYP_LONG);
+            assert(tree->gtType == TYP_INT);
+            assert((actualConstVal >= 0) && (actualConstVal <= INT32_MAX));
```
Comment (sandreenko) [inline src/jit/earlyprop.cpp]: Previously we did check value's range for `optPropKind::OPK_ARRAYLEN`, but not for `optPropKind::OPK_OBJ_GETTYPE`,  so I added this check.

File: src/jit/earlyprop.cpp [hunk 3]
```diff
@@ -228,9 +230,9 @@ void Compiler::optEarlyProp()
 //    tree           - The input tree node to be rewritten.
 //
 // Return Value:
-//    Return true iff "tree" is successfully rewritten.
+//    Return a new tree if the original tree was successfully rewritten.
```
Comment (CarolEidt) [inline src/jit/earlyprop.cpp]: You might also want to add a note that, if the tree is rewritten, its containing tree is updated (that wasn't obvious to me, given that previously the tree was modified in-place).

File: src/jit/earlyprop.cpp (hunk 1 above)
Comment (mikedn) [reply src/jit/earlyprop.cpp]: At the expense of adding new local variables and splitting the condition into multiple if statements. I don't think that such trivial expression can be considered "copy paste" to warrant this.

**Comment by sandreenko** [comment]
I found a problem with early replacement in call arguments, `replaceTree` doesn't update `fgArgTabEntry`, so it fails check in the lowering.

It was hidden because we did in-place replacement before. So I am checking the original code, to confirm that the replacement node size is always less or equal to the original node size (because `actualValCopy` is always constant int, that is small node), and we have never reached code, that called `replaceTree`, In this case I will add assert that `actualValClone` is small node and will call `CopyFrom`.

It is still necessary to return tree from this method, because it fixes `extra flag on a tree` for such cases:
```
Before optRemoveRangeCheck:
N008 (  4,  4) [000099] a---G--N----              /--*  IND       int
N006 (  1,  1) [001468] ------------              |  |  /--*  CNS_INT   long   16 Fseq[#FirstElem]
N007 (  2,  2) [001469] -------N----              |  \--*  ADD       byref
N005 (  1,  1) [001461] ------------              |     \--*  LCL_VAR   ref    V09 loc9         u:3
N009 ( 12, 15) [001470] ---XG--N----              *  COMMA     int
N004 (  8, 11) [001464] ---X--------              \--*  ARR_BOUNDS_CHECK_Rng void
N001 (  1,  1) [000097] ------------                 +--*  CNS_INT   int    0
N003 (  3,  3) [001463] ---X--------                 \--*  ARR_LENGTH int
N002 (  1,  1) [000096] ------------                    \--*  LCL_VAR   ref    V09 loc9         u:3
New refCnts for V09: refCnt = 55, refCntWtd = 55
After optRemoveRangeCheck:
N005 (  4,  4) [000099] a---G--N----              /--*  IND       int
N003 (  1,  1) [001468] ------------              |  |  /--*  CNS_INT   long   16 Fseq[#FirstElem]
N004 (  2,  2) [001469] -------N----              |  \--*  ADD       byref
N002 (  1,  1) [001461] ------------              |     \--*  LCL_VAR   ref    V09 loc9         u:3
N006 (  4,  4) [001470] ---XG--N----              *  COMMA     int <- EXTRA FLAG ON A TREE.
N001 (  0,  0) [002335] ------------              \--*  NOP       void
```

Because update side effects was called with the old tree, `001470` had wrong exception flag on it. We do not hit it, because usually we have exception flag on IND node also.

File: src/jit/earlyprop.cpp (hunk 1 above)
Comment (briansull) [reply src/jit/earlyprop.cpp]: This seems like a reasonable change to me.

File: src/jit/earlyprop.cpp [hunk 4]
```diff
@@ -228,9 +230,10 @@ void Compiler::optEarlyProp()
 //    tree           - The input tree node to be rewritten.
 //
 // Return Value:
-//    Return true iff "tree" is successfully rewritten.
-
-bool Compiler::optEarlyPropRewriteTree(GenTreePtr tree)
+//    Return a new tree if the original tree was successfully rewritten.
+//    The cointaining tree links are updated.
```
Comment (CarolEidt) [inline src/jit/earlyprop.cpp]: Minor: cointaining => containing

Comment (sandreenko) [reply src/jit/earlyprop.cpp]: Fixed, thanks.

**Pull request merged**